plugins/omdtls/omdtls.c

```diff
@@ -0,0 +1,885 @@
+	struct timeval timeout;
+	timeout.tv_sec = 3;
+	timeout.tv_usec = 0;
+	BIO_ctrl(bio_client, BIO_CTRL_DGRAM_SET_RECV_TIMEOUT, 0, &timeout);
+
+
+	pWrkrData->bIsConnected = 1;
+finalize_it:
+	DBGPRINTF("dtls_connect[%p]: Exit with %d for %s:%s\n", pWrkrData, iRet, pData->target, pData->port);
+	if(iRet != RS_RET_OK) {
+		pWrkrData->bIsConnected = 0;
+	}
+	RETiRet;
+}
+
+/*
+*	Setup socket and establish virtual Connect on a SOCKET Level for DTLS 
+*/
+static rsRetVal
+dtls_init(wrkrInstanceData_t *pWrkrData)
+{
+	DEFiRet;
+	int iErr;
+	instanceData *const pData = (instanceData *const) pWrkrData->pData;
+	struct addrinfo *addrResolved = NULL;
+	struct addrinfo hints;
+	DBGPRINTF("dtls_init[%p]: setup for %s:%s\n", pWrkrData, pData->target, pData->port);
+
+	// Close Before
+	dtls_close(pWrkrData);
+
+	// Setup receiving Socket for DTLS
+	if((pWrkrData->sockin = socket(AF_INET, SOCK_DGRAM, 0)) == -1)
+		return 1;
+	memset(&pWrkrData->dtls_client_addr, 0, sizeof(pWrkrData->dtls_client_addr));
+	pWrkrData->dtls_client_addr.sin_family = AF_INET;
+	pWrkrData->dtls_client_addr.sin_port = htons(0);
+	pWrkrData->dtls_client_addr.sin_addr.s_addr = INADDR_ANY;
+	if (bind(pWrkrData->sockin,
+		(struct sockaddr*)&pWrkrData->dtls_client_addr,
+		sizeof(pWrkrData->dtls_client_addr)) < 0) {
+		LogError(0, RS_RET_SUSPENDED, "omdtls[%p]: dtls_init Unable to bind DTLS CLient socket with errno %d",
+			pWrkrData, errno);
+		ABORT_FINALIZE(RS_RET_COULD_NOT_BIND);
+	}
+
+	// Resolve target address
+	memset(&hints, 0, sizeof(hints));
+	/* port must be numeric, because config file syntax requires this */
+	hints.ai_family = glbl.GetDefPFFamily(runModConf->pConf);
+	hints.ai_socktype = SOCK_DGRAM;
+	if((iErr = (getaddrinfo((char*)pData->target, (char*)pData->port, &hints, &addrResolved))) != 0) {
+		LogError(0, RS_RET_SUSPENDED,
+			"omdtls[%p]: could not get addrinfo for hostname '%s':'%s': %s",
+			pWrkrData, pData->target, pData->port, gai_strerror(iErr));
+		ABORT_FINALIZE(RS_RET_ADDRESS_UNKNOWN);
+	}
+	pWrkrData->dtls_rcvr_addrinfo = addrResolved;
+	addrResolved = NULL;
+
+	// Init Socket Connection (Which technically does not connect but prepares socket for DTLS)
+	DBGPRINTF("dtls_init[%p]: Init Session to %s:%s\n", pWrkrData, pData->target, pData->port);
+
+	pWrkrData->sockout = socket(AF_INET, SOCK_DGRAM, 0);
+	// Connect the UDP socket to the server's address
+	//	(const struct sockaddr *) &pWrkrData->dtls_rcvr_addr, sizeof(pWrkrData->dtls_rcvr_addr)) 
+	if (connect(pWrkrData->sockout,
+		pWrkrData->dtls_rcvr_addrinfo->ai_addr,
+		pWrkrData->dtls_rcvr_addrinfo->ai_addrlen) < 0) {
+		LogError(0, RS_RET_SUSPENDED,
+			"dtls_init[%p]: Failed to connect to hostname '%s':'%s': %s",
+			pWrkrData, pData->target, pData->port, gai_strerror(iErr));
+		ABORT_FINALIZE(RS_RET_ERR);
+	}
+	
+	// Socket Connect successfull, no continue with TLS
+	CHKiRet(dtls_connect(pWrkrData));
+
+	// We reached this position means we are ready to send!
+	pWrkrData->bIsConnecting = 0;
```

Comment:
do we need these two state variables?

I mean could we replace it wiht a single one with multiple values:

* disconnected
* connecting
* connected

I guess that these exact states must be gone through.

IIRC correctly, I suggested a similar replacement in some azure module, and you agreed to and implemented it.

Reply:
> Sure we can change intro a single variable with state values.

We should definitely do that - it's much cleaner and also will prevent hard-to-find errors if the current two vars for some reason get out of sync (by later maintenance).